src/io/flutter/utils/DiagnosticsNode.java

```diff
@@ -0,0 +1,502 @@
+   * describes.
+   *
+   * We cannot use observatory protocol InstanceRef ids as they are not adequately long lived.
+   */
+  private String dartDiagnosticId;
+
+  // XXX document
+  private String dartValueId;
+
+  private String getStringMember(String memberName) {
+    if (!json.has(memberName)) {
+      return null;
+    }
+    JsonElement value = json.get(memberName);
+    if (value instanceof JsonNull) {
+      return null;
+    }
+    return value.getAsString();
+  }
+
+  private boolean getBooleanMember(String memberName, boolean defaultValue) {
+    if (!json.has(memberName)) {
+      return defaultValue;
+    }
+    JsonElement value = json.get(memberName);
+    if (value instanceof JsonNull) {
+      return defaultValue;
+    }
+    return value.getAsBoolean();
+  }
+
+  private DiagnosticLevel getLevelMember(String memberName, DiagnosticLevel defaultValue) {
+    if (!json.has(memberName)) {
+      return defaultValue;
+    }
+    JsonElement value = json.get(memberName);
+    if (value instanceof JsonNull) {
+      return defaultValue;
+    }
+    return DiagnosticLevel.valueOf(value.getAsString());
+  }
+
+  private DiagnosticsTreeStyle getStyleMember(String memberName, DiagnosticsTreeStyle defaultValue) {
+    if (!json.has(memberName)) {
+      return defaultValue;
+    }
+    JsonElement value = json.get(memberName);
+    if (value instanceof JsonNull) {
+      return defaultValue;
+    }
+    return DiagnosticsTreeStyle.valueOf(value.getAsString());
+  }
+
+  /**
+   * Returns a reference to the value the DiagnosticsNode object is describing.
+   */
+   public InspectorInstanceRef getValueRef() {
+     final JsonElement valueId = json.get("valueId");
+    return new InspectorInstanceRef(valueId.isJsonNull() ? null : valueId.getAsString());
+  }
+
+  public boolean hasChildren() {
+    return getBooleanMember("hasChildren", false);
+  }
+
+  /**
+   * Check whether children are already available.
+   */
+  public boolean childrenReady() {
+    return children != null && children.isDone();
+  }
+
+  public CompletableFuture<ArrayList<DiagnosticsNode>> getChildren() {
+    if (children == null) {
+      if (hasChildren()) {
+        children = inspectorService.getChildren(getDartDiagnosticRef());
+      } else {
+        // Known to have no children so we can provide the children immediately.
+        children = new CompletableFuture<ArrayList<DiagnosticsNode>>();
+        children.complete(new ArrayList<>());
+      }
+    }
+    return children;
+  }
+
+  /**
+   * Reference the actual Dart DiagnosticsNode object this object is referencing.
+   */
+  public InspectorInstanceRef getDartDiagnosticRef() {
+    return new InspectorInstanceRef(json.get("objectId").getAsString());
+  }
+
+  public CompletableFuture<ArrayList<DiagnosticsNode>> getProperties() {
+    if (properties == null) {
+      properties = inspectorService.getProperties(getDartDiagnosticRef());
+    }
+    return properties;
+  }
+}
```

Comment:
just to reduce future diffs, I would do a formatting pass on the files

Reply:
done